### DISCUSSION TITLE
Gitea: Personal repos not being shown?

### DISCUSSION THREAD

**Discussion by pat-s** [woodpecker-ci/woodpecker#2491] [Ask a Question]
Forge: Gitea

So far I only worked with repos in orgs.
I tried to enable a private repo now but it doesn't come up when clicking "add repository" which essentially just invokes `/api/user/repos` AFAIK.
I logged in and out again, changed the log level to "debug" but still can't infer why it doesn't show up. 

I don't have any restrictions set on the server in `WOODPECKER_ORGS` or similar. And I'd suspect that personal repos should always appear for the respective logged-in user?

This is a bit embarrassing but I might be missing something obvious here?

`next-baaf8b97e1`

**Comment by qwerty287** [comment]
What do you mean by "private" repo? A repo owned by yourself (and not an org) or really a private repo with limited visibility?

**Reply by pat-s** [reply]
A repo owned by myself.

**Reply by qwerty287** [reply]
Weird, that's totally working for me. Can you check the `/api/user/repos` endpoint?

**Reply by pat-s** [reply]
It works there and strangely it works now also in the UI.

I swear I tried like 10 times yesterday and it didn't show up continuously over a period of a few hours. Strange.

Let's book it under "strange things that can't be explained" 😄

**Reply by pat-s** [reply]
Ups, checked with the wrong forge. 

-> GH works
-> Gitea doesn't

and `/api/user/repos` via swagger doesn't show the personal repos in my case. Only org repos.

**Reply by pat-s** [reply]
Can you recreate this with your instance? I am now on `next-f8e91f00aa` and apparently `/api/user/repos` doesn't show my personal repos, only the one from orgs I am a member of.

**Reply by qwerty287** [reply]
It was working for me when I tried yesterday. I'm using this all the time on my dev instance...

**Reply by pat-s** [reply]
Hm. I mean it can only be related to permissions, either of the token or the WP application which is required for Gitea. I'll do some more digging in the next days.

**Reply by pat-s** [reply]
I have another instance of this strange behavior.

- New org in Gitea
- Two repos (same access, permissions, etc.)
- Only one shows up in WP after syncing

Login/loguout, still after 15 mins only one of the two shows up when clicking "add repo" 🤔

**Reply by qwerty287** [reply]
Can you use a custom woodpecker binary and add some debug log statements in the gitea driver to check whether the repos are existing here? If not it's either a gitea or a gitea sdk issue, otherwise ours.

**Reply by pat-s** [reply]
Logging my actions as meanwhile I can't add any new repo anymore (next (Nov 02), Gitea 1.20.5):

- I realized that I got error when "repairing all repos" in the admin interface
	- Disabled the problematic ones and made sure to get the call working
- Afters I noticed the following ones in the logs when using "add repository":

```
{"level":"error","time":"2023-11-02T11:43:00Z","caller":"/woodpecker/src/github.com/woodpecker-ci/woodpecker/server/router/middleware/session/repo.go:115","message":"Error fetching permission for patrick . sql: no rows in result set"}
```

In addition I see 

```
 {"level":"debug","time":"2023-11-02T11:43:00Z","caller":"/woodpecker/src/github.com/woodpecker-ci/woodpecker/server/router/middleware/session/repo.go:150","message":"patrick granted &{UserID:0 RepoID:0 Repo │
│ :<nil> Pull:true Push:true Admin:true Synced:0 Created:0 Updated:0} permission to "}
```

which does not look like a successful execution (?).

The `sql: no rows in result set` error was not appearing before I fixed the "repair all repos" call. This might be related to https://github.com/woodpecker-ci/woodpecker/issues/1851

I also came across https://github.com/woodpecker-ci/woodpecker/issues/1826 which might be related.

**Reply by pat-s** [reply]
I meanwhile suspect that something is off with the DB after going through all the updates. I remember having reverted to 1.0.2 from next in between which have distorted things in the DB?
It might be that using a new/fresh DB is the only/best way forward?

**Reply by pat-s** [reply]
Using a new DB did not help :/

The strange thing which I don't understand: Some repos of an org are shown, some not. Both are similar (i.e. same permissions, etc.). My user is even an admin of the org 🤔

**Reply by pat-s** [reply]
I think it might be https://github.com/woodpecker-ci/woodpecker/issues/1826. 

OTOH, I've created a new org + repo with my user and it also doesn't show up. No other user involved.
What I am wondering a bit is why nobody else has reported similar issues lately (if it's a generic issue). The WP-next + Gitea combo should be the most used one 🤔

**Reply by pat-s** [reply]
@qwerty287 

> Can you use a custom woodpecker binary and add some debug log statements in the gitea driver to check whether the repos are existing here? If not it's either a gitea or a gitea sdk issue, otherwise ours.

This might be needed at this point I guess. However, I am not sure my go skills are enough to get something helpful going. Would this function be a good target? https://github.com/woodpecker-ci/woodpecker/blob/f270be49403c53bf10e949a854b432f798dc09a5/server/forge/gitea/gitea.go#L254-L280

**Reply by qwerty287** [reply]
> I think it might be #1826.

No, because this issue refers to repos that are activated by a different user - you can't even activate it.

> The WP-next + Gitea combo should be the most used one 🤔

It's running totally fine as my dev setup.

> Would this function be a good target?

Yes, this one and the one where it's called (the api handler for listing repos).

**Reply by mzampetakis** [reply]
At first I was suspicious about the DB, but then I realised that you are not even enabling a repo. So, tried it in my Gitea and WP at #7bc40f20cb3edb4aae3effe451792e736b9f9cf6 (local dev), and I successfully got them all. I have 2 orgs with different access level and 2 personal repos:

![gitea](https://github.com/woodpecker-ci/woodpecker/assets/16765483/458cf71c-b2fb-428e-bbb8-aa4659b81fe7)

Have you checked if any repo is archived? Looking at the code https://github.com/woodpecker-ci/woodpecker/discussions/2491#discussioncomment-7456243 I cannot see anything suspicious.

**Reply by pat-s** [reply]
> Have you checked if any repo is archived? Looking at the code https://github.com/woodpecker-ci/woodpecker/discussions/2491#discussioncomment-7456243 I cannot see anything suspicious.

Not archived. Even switched through "private/internal/public" without success. 
Tried a brand new org with new repos, no success.

For some orgs (which I created myself), I see some repos but others not. Also none which I create fresh.

-> Something must be off with the syncing from Gitea (API only gives 200 as return but it doesn't return all repos I have access to.

> At first I was suspicious about the DB, but then I realised that you are not even enabling a repo

Yes, I would like to enable them but they are not shown in the available repos which are coming from the Gitea sync.

**Reply by qwerty287** [reply]
> Something must be off with the syncing from Gitea (API only gives 200 as return but it doesn't return all repos I have access to.

So the gitea API does not return all repos, only those that are displayed in woodpecker? Or do you mean woodpecker's API?

**Reply by pat-s** [reply]
I **assume** that the Gitea SDK within WP does not return all repos (which is why I think #2705 might help but I can't build the PR images atm :/). Haven't yet checked with the Gitea API directly, but good point, I should check that next.

**Reply by mzampetakis** [reply]
You could also try to run a local woodpecker server and debug through appropriate logging. Alternatively you could check [Gitea's API](https://docs.gitea.com/development/api-usage) directly for its response.

**Reply by pat-s** [reply]
So, here we go: `user/repos` of the Gitea API does not return the questionable repos as well.
Super strange, as I created them and can push to them directly 😄 
Gitea 1.20.5

So good news: it seems not be a WP issue and also not an issue of the Gitea SDK.

**Reply by mzampetakis** [reply]
Great finding! So if you find or open any relevant Gitea issue please mention it here for the case that someone else faces the same problem.

**Comment by qwerty287** [comment]
So this has been reported upstream: https://github.com/go-gitea/gitea/issues/27891

I'm closing this one as we can't fix it.

**Comment by pat-s** [comment]
I meanwhile found out that it's a Gitea pagination issue. I can retrieve the repos via the Gitea UI when applying some pagination. By default, only the first 50 repos seem to be returned.

The following issues will explain this: https://gitea.com/gitea/go-sdk/issues/507

Maybe we incorporated the issues with the gitea skd upgrade in https://github.com/woodpecker-ci/woodpecker/pull/2464. There is no changelog for 0.16 but a lot of commits and some which tackle pagination like https://gitea.com/gitea/go-sdk/pulls/573. Maybe changes are needed in the WP WRT to Gitea pagination.

The issue might only appear if a user has access to a lot of repos (e.g. when being the admin of a large instance, like it is the case in the references case of mine).

**Reply by qwerty287** [reply]
Actually we always fetch all repos 🤔

https://github.com/woodpecker-ci/woodpecker/blob/6ab0e800dc1f76178943126ca8503edda25c1cb7/server/forge/gitea/gitea.go#L262

**Reply by pat-s** [reply]
I can't infer the exact actions from looking at the code but after changing `MAX_RESPONSE_ITEMS` and `DEFAULT_PAGING_NUM` to 10000 in Gitea, I can get all repos in a single API call without pagination. However, they still don't show up in WP after a sync. Likely because WP enforces the paginated approach which might need an update after updating go-gitea to 0.16? Just a guess for now...

**Reply by pat-s** [reply]
I'll build an image with go-gitea 0.15.1 and see if that resolves the issues to verify the claim.

**Reply by pat-s** [reply]
I can confirm that building an image with an old server binary which runs with go-gitea 0.15.1 resolves the issue. So the issue is related to the 0.16 upgrade.

**Reply by pat-s** [reply]
Or maybe, looking at the commit log, it's https://github.com/woodpecker-ci/woodpecker/pull/2633. I'll check the image right before this one as the go-gitea SDK update was on September 16th and I suspect a pagination issue.

**Reply by pat-s** [reply]
So in `7b3eeec65c` (Oct 17th) it's already broken. Testing now the image right before the gitea sdk upgrade.

**Reply by pat-s** [reply]
Hmm 9433c739e34cda052bcabc05619021e16ee91762 (Sep 13th) is also not working 🤔

**Reply by pat-s** [reply]
NB: the server binary I used for the image I've built myself is from May 17th (it was still on my local disk). Long way to go...

**Reply by pat-s** [reply]
Narrowed it down between Sept 3 (working) and Sept 8 (failing). There are 7 commits in this range with most of them being docs-only, build-only or translations. The main ones in question are 

- 85e157bb10da2c1142251824e0f79a33e719bb7a
- 8a517a8596e11b87f5c54baf1d55d113fd3e3409

**Reply by qwerty287** [reply]
Yes that's it. It's the second one you listed there.

Reason: Probably you have some archived repos? If no, the commit is not the issue (would be really weird then though). If yes: The pagination utils must receive the _full_ response as they compare it to the len of the first response. If it's smaller (-> last page), it finishes. If some repos are filtered out, this happens already on a page that's not the last one because the len of the returned list is smaller than the first response.

I can send fix for this, but I don't know when I get the time for this (I'm writing this from my phone...).

**Reply by pat-s** [reply]
Yes that makes sense and also matches my debugging results: I tracked it down to https://github.com/woodpecker-ci/woodpecker/pull/2374 and the 

```
			if repo.GetArchived() {
				continue
			}
```

line. Removing it, makes things work again. I can send a quickfix-revert for now until a proper fix is ready.

**Reply by pat-s** [reply]
What a relief! I was already getting crazy about this one and it blocked our operations quite a bit as no new repos could be added 😅

**Reply by qwerty287** [reply]
The proper fix would just be to move this outside of the pagination func and iterate over all items in the final list instead and only keep the non-archived
